File: part02/L1425.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.stream.IntStream;

public class L1425 {
    public static void addSub(String add, String sub, HashMap<String, Integer> count) {
        if (add != null)
            count.put(add, count.getOrDefault(add, 0) + 1);
        if (sub != null) {
            int tmp = count.get(sub);
            if (tmp > 1)
                count.put(sub, tmp - 1);
            else
                count.remove(sub);
        }
    }

    public static ArrayList<Integer> solution(String[] gems) {
        HashSet<String> set = new HashSet<String>(Arrays.asList(gems));
        HashMap<String, Integer> count = new HashMap<String, Integer>();
        addSub(gems[0], null, count);
        int numTypes = set.size(), last = gems.length;
        int l = 0, r = 0, sl = 1, sr = last;
        while (l <= r && r < last) {
            if (count.size() == numTypes) {
                if (r - l < sr - sl) {
                    sl = l + 1;
                    sr = r + 1;
                }
                addSub(null, gems[l], count);
                l++;
                continue;
            }
            r++;
            if (r < last)
                addSub(gems[r], null, count);
        }
        return IntStream.of(sl, sr).collect(ArrayList::new, ArrayList::add, ArrayList::addAll);
    }

    public static void main(String[] args) {
        // Test code
        String[] gems = {"DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"};
        System.out.println(solution(gems));

        gems = new String[]{"AA", "AB", "AC", "AA", "AC"};
        System.out.println(solution(gems));

        gems = new String[]{"XYZ", "XYZ", "XYZ"};
        System.out.println(solution(gems));

        gems = new String[]{"ZZZ", "YYY", "NNNN", "YYY", "BBB"};
        System.out.println(solution(gems));
    }
}

/* programmers submission
    // Source: [카카오 인턴] 보석 쇼핑 https://programmers.co.kr/learn/courses/30/lessons/67258
    // Submission detail:
    //     정확성  테스트
    //     테스트 1 〉	통과 (0.16ms, 84.1MB)
    //     테스트 2 〉	통과 (0.26ms, 73.6MB)
    //     테스트 3 〉	통과 (0.61ms, 78.6MB)
    //     테스트 4 〉	통과 (0.53ms, 75.6MB)
    //     테스트 5 〉	통과 (0.20ms, 76.6MB)
    //     테스트 6 〉	통과 (0.17ms, 76.1MB)
    //     테스트 7 〉	통과 (0.15ms, 73MB)
    //     테스트 8 〉	통과 (0.87ms, 69.2MB)
    //     테스트 9 〉	통과 (0.97ms, 87.1MB)
    //     테스트 10 〉	통과 (0.97ms, 76.8MB)
    //     테스트 11 〉	통과 (0.92ms, 76.3MB)
    //     테스트 12 〉	통과 (1.47ms, 87.4MB)
    //     테스트 13 〉	통과 (1.52ms, 78.5MB)
    //     테스트 14 〉	통과 (1.48ms, 82.5MB)
    //     테스트 15 〉	통과 (2.17ms, 80.3MB)
    //     효율성  테스트
    //     테스트 1 〉	통과 (6.90ms, 55.2MB)
    //     테스트 2 〉	통과 (7.13ms, 56.9MB)
    //     테스트 3 〉	통과 (14.31ms, 58.4MB)
    //     테스트 4 〉	통과 (10.00ms, 60.1MB)
    //     테스트 5 〉	통과 (16.16ms, 62.3MB)
    //     테스트 6 〉	통과 (17.59ms, 64.6MB)
    //     테스트 7 〉	통과 (26.76ms, 64.5MB)
    //     테스트 8 〉	통과 (24.75ms, 67.6MB)
    //     테스트 9 〉	통과 (35.58ms, 70.4MB)
    //     테스트 10 〉	통과 (35.93ms, 74.7MB)
    //     테스트 11 〉	통과 (49.58ms, 79.2MB)
    //     테스트 12 〉	통과 (43.07ms, 80MB)
    //     테스트 13 〉	통과 (61.16ms, 79.9MB)
    //     테스트 14 〉	통과 (52.36ms, 97.8MB)
    //     테스트 15 〉	통과 (60.42ms, 81.2MB)
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public static void addSub(String add, String sub, HashMap<String, Integer> count) {
        if (add != null)
            count.put(add, count.getOrDefault(add, 0) + 1);
        if (sub != null) {
            int tmp = count.get(sub);
            if (tmp > 1)
                count.put(sub, tmp - 1);
            else
                count.remove(sub);
        }
    }

    public int[] solution(String[] gems) {
        HashSet<String> set = new HashSet<String>(Arrays.asList(gems));
        HashMap<String, Integer> count = new HashMap<String, Integer>();
        addSub(gems[0], null, count);
        int numTypes = set.size(), last = gems.length;
        int l = 0, r = 0, sl = 1, sr = last;
        while (l <= r && r < last) {
            if (count.size() == numTypes) {
                if (r - l < sr - sl) {
                    sl = l + 1;
                    sr = r + 1;
                }
                addSub(null, gems[l], count);
                l++;
                continue;
            }
            r++;
            if (r < last)
                addSub(gems[r], null, count);
        }
        return new int[] {sl, sr};
    }
}
*/

File: part02/L1425.py
from time import perf_counter_ns as ns


def solution(gems):
    d, num_types = {gems[0]: 1}, len(set(gems))
    ans = [1, (length := len(gems))]
    l = r = 0
    while l <= r < length:
        if len(d) == num_types:
            ans = [l + 1, r + 1] if ans[1] - ans[0] > r - l else ans
            d[gems[l]] -= 1
            if not d[gems[l]]:
                del d[gems[l]]
            l += 1
            continue
        r += 1
        if r < length:
            d[gems[r]] = d.get(gems[r], 0) + 1
    return ans


if __name__ == '__main__':
    ITERATIONS = 10_000
    print(f'Running the basic tests {ITERATIONS:,} times...')
    tests = (
        (["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"], [3, 7]),
        (["AA", "AB", "AC", "AA", "AC"], [1, 3]),
        (["XYZ", "XYZ", "XYZ"], [1, 1]),
        (["ZZZ", "YYY", "NNNN", "YYY", "BBB"], [1, 5])
    )
    for gems, expected in tests:
        print(f'solution({gems}) returned', end=' ')
        if (result := solution(gems)) == expected:
            print(f'the expected result {expected}', end=' ')
            fastest = float('inf')
            slowest = total = 0
            for _ in range(ITERATIONS):
                start = ns()
                solution(gems)
                end = ns()
                time = end - start
                fastest, slowest = min(time, fastest), max(time, slowest)
                total += time
            print(f'in an average of {total / ITERATIONS / 1000:,.2f}ms '
                  f'(min: {fastest / 1000:,.2f}ms, '
                  f'max: {slowest / 1000:,.2f}ms)')
        else:
            print(f'a wrong result {result} (expected: {expected})')


''' programmers submission
    # Source: [카카오 인턴] 보석 쇼핑 https://programmers.co.kr/learn/courses/30/lessons/67258
    # Submission detail:
    #     정확성  테스트
    #     테스트 1 〉	통과 (0.02ms, 10.1MB)
    #     테스트 2 〉	통과 (0.06ms, 10.2MB)
    #     테스트 3 〉	통과 (0.33ms, 10.3MB)
    #     테스트 4 〉	통과 (0.18ms, 10.1MB)
    #     테스트 5 〉	통과 (0.02ms, 10.1MB)
    #     테스트 6 〉	통과 (0.01ms, 10.2MB)
    #     테스트 7 〉	통과 (0.02ms, 10.4MB)
    #     테스트 8 〉	통과 (0.33ms, 10.1MB)
    #     테스트 9 〉	통과 (0.55ms, 10.1MB)
    #     테스트 10 〉	통과 (0.33ms, 10.3MB)
    #     테스트 11 〉	통과 (0.49ms, 10.3MB)
    #     테스트 12 〉	통과 (0.90ms, 10.2MB)
    #     테스트 13 〉	통과 (1.18ms, 10.4MB)
    #     테스트 14 〉	통과 (1.02ms, 10.4MB)
    #     테스트 15 〉	통과 (2.46ms, 10.3MB)
    #     효율성  테스트
    #     테스트 1 〉	통과 (3.12ms, 10.4MB)
    #     테스트 2 〉	통과 (4.80ms, 10.6MB)
    #     테스트 3 〉	통과 (10.07ms, 11.1MB)
    #     테스트 4 〉	통과 (7.46ms, 11.8MB)
    #     테스트 5 〉	통과 (16.57ms, 11.8MB)
    #     테스트 6 〉	통과 (20.17ms, 12.1MB)
    #     테스트 7 〉	통과 (23.82ms, 12.6MB)
    #     테스트 8 〉	통과 (26.01ms, 12.9MB)
    #     테스트 9 〉	통과 (30.12ms, 13.3MB)
    #     테스트 10 〉	통과 (35.76ms, 13.7MB)
    #     테스트 11 〉	통과 (39.25ms, 14.7MB)
    #     테스트 12 〉	통과 (26.71ms, 15.4MB)
    #     테스트 13 〉	통과 (38.51ms, 16.2MB)
    #     테스트 14 〉	통과 (62.15ms, 17MB)
    #     테스트 15 〉	통과 (62.64ms, 17.7MB)
def solution(gems):
    d, num_types = {gems[0]: 1}, len(set(gems))
    ans = [1, (length := len(gems))]
    l = r = 0
    while l <= r < length:
        if len(d) == num_types:
            ans = [l + 1, r + 1] if ans[1] - ans[0] > r - l else ans
            d[gems[l]] -= 1
            if not d[gems[l]]:
                del d[gems[l]]
            l += 1
            continue
        r += 1
        if r < length:
            d[gems[r]] = d.get(gems[r], 0) + 1
    return ans
'''
